File: inventario_python/dao/cliente_dao.py
from db.conexion import conectar
from models.cliente import Cliente
from utils.aes_util import encriptar, desencriptar


class ClienteDAO:

    def _mapear(self, row):
        c = Cliente()
        c.id = row[0]
        c.cedula = desencriptar(row[1]) if row[1] else ""
        c.nombre = desencriptar(row[2]) if row[2] else ""
        c.correo = desencriptar(row[3]) if row[3] else ""
        c.telefono = desencriptar(row[4]) if row[4] else ""
        return c

    def listar_todos(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute("SELECT * FROM cliente ORDER BY id DESC")
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def buscar(self, texto):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            enc = encriptar(texto)
            cur.execute("SELECT * FROM cliente WHERE nombre LIKE %s OR cedula LIKE %s OR nombre = %s OR cedula = %s ORDER BY id DESC", (f"%{texto}%", f"%{texto}%", enc, enc))
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def insertar(self, c):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("INSERT INTO cliente (cedula, nombre, correo, telefono) VALUES (%s,%s,%s,%s)",
                        (encriptar(c.cedula) if c.cedula else None, encriptar(c.nombre),
                         encriptar(c.correo) if c.correo else None, encriptar(c.telefono) if c.telefono else None))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def actualizar(self, c):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("UPDATE cliente SET cedula=%s, nombre=%s, correo=%s, telefono=%s WHERE id=%s",
                        (encriptar(c.cedula) if c.cedula else None, encriptar(c.nombre),
                         encriptar(c.correo) if c.correo else None, encriptar(c.telefono) if c.telefono else None, c.id))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def eliminar(self, id):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("DELETE FROM cliente WHERE id = %s", (id,))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def contar_todos(self):
        conn = conectar()
        if not conn:
            return 0
        try:
            cur = conn.cursor()
            cur.execute("SELECT COUNT(*) FROM cliente")
            return cur.fetchone()[0]
        finally:
            conn.close()

File: inventario/src/DAO/ClienteDAO.java
package DAO;

import CX.ConexionBD;
import Modelo.Cliente;
import Utils.AESUtil;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class ClienteDAO {

    public List<Cliente> listarTodos() throws SQLException {
        List<Cliente> lista = new ArrayList<>();
        String sql = "SELECT * FROM cliente ORDER BY id DESC";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) lista.add(mapear(rs));
        }
        return lista;
    }

    public List<Cliente> buscar(String texto) throws SQLException {
        List<Cliente> lista = new ArrayList<>();
        String enc = AESUtil.encriptar(texto);
        String sql = "SELECT * FROM cliente WHERE nombre LIKE ? OR cedula LIKE ? OR nombre = ? OR cedula = ? ORDER BY id DESC";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setString(1, "%" + texto + "%");
            ps.setString(2, "%" + texto + "%");
            ps.setString(3, enc);
            ps.setString(4, enc);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) lista.add(mapear(rs));
            }
        }
        return lista;
    }

    public void insertar(Cliente c) throws SQLException {
        String sql = "INSERT INTO cliente (cedula, nombre, correo, telefono) VALUES (?, ?, ?, ?)";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setString(1, AESUtil.encriptar(c.getCedula()));
            ps.setString(2, AESUtil.encriptar(c.getNombre()));
            ps.setString(3, AESUtil.encriptar(c.getCorreo()));
            ps.setString(4, AESUtil.encriptar(c.getTelefono()));
            ps.executeUpdate();
        }
    }

    public void actualizar(Cliente c) throws SQLException {
        String sql = "UPDATE cliente SET cedula=?, nombre=?, correo=?, telefono=? WHERE id=?";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setString(1, AESUtil.encriptar(c.getCedula()));
            ps.setString(2, AESUtil.encriptar(c.getNombre()));
            ps.setString(3, AESUtil.encriptar(c.getCorreo()));
            ps.setString(4, AESUtil.encriptar(c.getTelefono()));
            ps.setLong(5, c.getId());
            ps.executeUpdate();
        }
    }

    public void eliminar(long id) throws SQLException {
        String sql = "DELETE FROM cliente WHERE id=?";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setLong(1, id);
            ps.executeUpdate();
        }
    }

    public int contarTotal() throws SQLException {
        String sql = "SELECT COUNT(*) FROM cliente";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {
            if (rs.next()) return rs.getInt(1);
        }
        return 0;
    }

    private Cliente mapear(ResultSet rs) throws SQLException {
        Cliente c = new Cliente();
        c.setId(rs.getLong("id"));
        c.setCedula(AESUtil.desencriptar(rs.getString("cedula")));
        c.setNombre(AESUtil.desencriptar(rs.getString("nombre")));
        c.setCorreo(AESUtil.desencriptar(rs.getString("correo")));
        c.setTelefono(AESUtil.desencriptar(rs.getString("telefono")));
        return c;
    }
}
